comps/knowledgegraphs/langchain/knowledge_graph.py

```diff
@@ -0,0 +1,161 @@
+        node_label=input.rag_node_label,
+        text_node_properties=input.rag_text_node_properties,
+        embedding_node_property=input.rag_embedding_node_property,
+    )
+    vector_qa = RetrievalQA.from_chain_type(llm=llm, chain_type="stuff", retriever=vector_index.as_retriever())
+    return vector_qa
+
+
+def get_cypherchain(graph, cypher_llm, qa_llm):
+    graph.refresh_schema()
+    cypher_chain = GraphCypherQAChain.from_llm(cypher_llm=cypher_llm, qa_llm=qa_llm, graph=graph, verbose=True)
+    return cypher_chain
+
+
+def get_agent(vector_qa, cypher_chain, llm_repo_id):
+    # define two tools
+    tools = [
+        Tool(
+            name="Tasks",
+            func=vector_qa.invoke,
+            description="""Useful when you need to answer questions about descriptions of tasks.
+            Not useful for counting the number of tasks.
+            Use full question as input.
+            """,
+        ),
+        Tool(
+            name="Graph",
+            func=cypher_chain.invoke,
+            description="""Useful when you need to answer questions about microservices,
+            their dependencies or assigned people. Also useful for any sort of
+            aggregation like counting the number of tasks, etc.
+            Use full question as input.
+            """,
+        ),
+    ]
+
+    # setup ReAct style prompt
+    prompt = hub.pull("hwchase17/react-json")
+    prompt = prompt.partial(
+        tools=render_text_description(tools),
+        tool_names=", ".join([t.name for t in tools]),
+    )
+
+    # define chat model
+    llm = HuggingFaceEndpoint(repo_id=llm_repo_id, max_new_tokens=512)
+    chat_model = ChatHuggingFace(llm=llm)
+    chat_model_with_stop = chat_model.bind(stop=["\nObservation"])
+
+    # define agent
+    agent = (
+        {
+            "input": lambda x: x["input"],
+            "agent_scratchpad": lambda x: format_log_to_str(x["intermediate_steps"]),
+        }
+        | prompt
+        | chat_model_with_stop
+        | ReActJsonSingleInputOutputParser()
+    )
+
+    # instantiate AgentExecutor
+    agent_executor = AgentExecutor(agent=agent, tools=tools, verbose=True)
+    return agent_executor
+
+
+@register_microservice(
+    name="opea_service@knowledge_graph",
+    endpoint="/v1/graphs",
+    host="0.0.0.0",
+    port=8060,
+)
+def graph_query(input: GraphDoc) -> GeneratedDoc:
+    print(input)
+
+    ## Connect to Neo4j
+    neo4j_endpoint = os.getenv("NEO4J_ENDPOINT", "neo4j://localhost:7687")
+    neo4j_username = os.getenv("NEO4J_USERNAME", "neo4j")
+    neo4j_password = os.getenv("NEO4J_PASSWORD", "neo4j")
+    graph = Neo4jGraph(url=neo4j_endpoint, username=neo4j_username, password=neo4j_password)
+
+    ## keep for multiple tests, will remove later
+    graph.query("MATCH (n) DETACH DELETE n")
+    import_url = "https://gist.githubusercontent.com/tomasonjo/08dc8ba0e19d592c4c3cde40dd6abcc3/raw/e90b0c9386bf8be15b199e8ac8f83fc265a2ac57/microservices.json"
```

Comment:
What is this url used for?

Reply:
It's original for downloading example json data. Have refined this, remove the url and add the example data to comps/knowledgegraphslangchain/data/microservice.json